crypto/evp/mac_lib.c

```diff
@@ -52,7 +52,8 @@ int EVP_MAC_CTX_copy(EVP_MAC_CTX *dst, EVP_MAC_CTX *src)
 {
     EVP_MAC_IMPL *macdata;
 
-    if (src->data != NULL && !dst->meth->copy(dst->data, src->data))
+    if (src->meth != dst->meth
```

Comment:
Shouldn't copy replace the method too?

Reply:
Err.. They are different structure with different sizes, so it would need to rellocate the dst data block size somehow before it could think about doing that.

Reply:
I guess the question is now should reallocate what's necessary?
Thoughts @openssl/committers ?

Reply:
One way to make progress on such questions is perhaps to write documentation that explains the expected semantics.  That usually forces one to focus on the design away from the distractions of the code.

Reply:
@vdukhovni good point.  Thanks.

Reply:
I think the other way, that you shouldn't try to copy contexts between different implementations.  Doing so leads to undefined results.  So this test seems correct to me.

Reply:
The following code seems to work. Does anyone have a strong opinion on which is better?

    /* If the mac's are different, remove the old implementation and replace */
    if (src->meth != dst->meth) {
        dst->meth->free(dst->data);
        dst->meth = src->meth;
        dst->data = src->meth->new();
        if (dst->data == NULL)
            return 0;
    }
    return src->meth->copy(dst->data, src->data);

Reply:
In #7561 I've replaced the copy method with a dup method, which avoids this whole thing.

Reply:
Ok. Maybe I should revisit this PR after your changes are merged. The initial problem I was trying to solve was making repeated calls to init/update/final work correctly. The copy method then tied into this..

Reply:
A dup method sounds better.